File: packages/agent/services/web-search.ts
import type { FetchFn } from '../types'

export interface WebSearchResult {
  title: string
  url: string
  snippet: string
}

export interface WebSearchConfig {
  enabled: boolean
  provider: string
  apiKey?: string
}

export async function webSearch(
  fetchFn: FetchFn,
  config: WebSearchConfig,
  query: string,
  limit = 8,
): Promise<WebSearchResult[]> {
  if (!config.enabled || !query.trim()) return []

  const q = query.trim()
  switch (config.provider) {
    case 'serper':
      return searchSerper(fetchFn, config.apiKey ?? '', q, limit)
    case 'tavily':
      return searchTavily(fetchFn, config.apiKey ?? '', q, limit)
    default:
      return searchDuckDuckGo(fetchFn, q, limit)
  }
}

async function searchSerper(
  fetchFn: FetchFn,
  apiKey: string,
  query: string,
  limit: number,
): Promise<WebSearchResult[]> {
  if (!apiKey) throw new Error('Serper 需要 API Key')
  const res = await fetchFn('https://google.serper.dev/search', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-API-KEY': apiKey,
    },
    body: JSON.stringify({ q: query, num: limit }),
  })
  if (!res.ok) {
    const body = await res.text()
    throw new Error(`Serper 搜索失败 (${res.status}): ${body.slice(0, 200)}`)
  }
  const data = (await res.json()) as {
    organic?: Array<{ title?: string; link?: string; snippet?: string }>
  }
  return (data.organic ?? []).slice(0, limit).map((item) => ({
    title: item.title ?? '',
    url: item.link ?? '',
    snippet: item.snippet ?? '',
  }))
}

async function searchTavily(
  fetchFn: FetchFn,
  apiKey: string,
  query: string,
  limit: number,
): Promise<WebSearchResult[]> {
  if (!apiKey) throw new Error('Tavily 需要 API Key')
  const res = await fetchFn('https://api.tavily.com/search', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      api_key: apiKey,
      query,
      max_results: limit,
      search_depth: 'basic',
    }),
  })
  if (!res.ok) {
    const body = await res.text()
    throw new Error(`Tavily 搜索失败 (${res.status}): ${body.slice(0, 200)}`)
  }
  const data = (await res.json()) as {
    results?: Array<{ title?: string; url?: string; content?: string }>
  }
  return (data.results ?? []).slice(0, limit).map((item) => ({
    title: item.title ?? '',
    url: item.url ?? '',
    snippet: item.content ?? '',
  }))
}

async function searchDuckDuckGo(
  fetchFn: FetchFn,
  query: string,
  limit: number,
): Promise<WebSearchResult[]> {
  const url = `https://api.duckduckgo.com/?q=${encodeURIComponent(query)}&format=json&no_html=1&skip_disambig=1`
  const res = await fetchFn(url)
  if (!res.ok) return []
  const data = (await res.json()) as {
    AbstractText?: string
    AbstractURL?: string
    Heading?: string
    RelatedTopics?: Array<{ Text?: string; FirstURL?: string; Topics?: Array<{ Text?: string; FirstURL?: string }> }>
  }
  const results: WebSearchResult[] = []
  if (data.AbstractText && data.AbstractURL) {
    results.push({
      title: data.Heading ?? query,
      url: data.AbstractURL,
      snippet: data.AbstractText,
    })
  }
  for (const topic of data.RelatedTopics ?? []) {
    if (topic.Text && topic.FirstURL) {
      results.push({ title: topic.Text.slice(0, 80), url: topic.FirstURL, snippet: topic.Text })
    }
    for (const sub of topic.Topics ?? []) {
      if (sub.Text && sub.FirstURL) {
        results.push({ title: sub.Text.slice(0, 80), url: sub.FirstURL, snippet: sub.Text })
      }
    }
    if (results.length >= limit) break
  }
  return results.slice(0, limit)
}

File: src-tauri/src/web_search.rs
use crate::config::load_config;
use serde::Serialize;
use tauri::AppHandle;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct WebSearchItem {
    pub title: String,
    pub url: String,
    pub snippet: String,
}

#[tauri::command]
pub async fn web_search(
    app: AppHandle,
    query: String,
    limit: Option<u32>,
) -> Result<Vec<WebSearchItem>, String> {
    let config = load_config(&app)?;
    if !config.search.enabled {
        return Ok(vec![]);
    }
    let limit = limit.unwrap_or(8).min(20) as usize;
    let q = query.trim();
    if q.is_empty() {
        return Ok(vec![]);
    }

    match config.search.provider.as_str() {
        "serper" => search_serper(&config.search.api_key, q, limit).await,
        "tavily" => search_tavily(&config.search.api_key, q, limit).await,
        _ => search_duckduckgo(q, limit).await,
    }
}

async fn search_serper(api_key: &str, query: &str, limit: usize) -> Result<Vec<WebSearchItem>, String> {
    if api_key.is_empty() {
        return Err("Serper 需要 API Key".into());
    }
    let body = serde_json::json!({ "q": query, "num": limit });
    let response = crate::proxy::send_http(
        "POST",
        "https://google.serper.dev/search",
        Some(&[
            ("Content-Type", "application/json"),
            ("X-API-KEY", api_key),
        ]),
        Some(body.to_string()),
        Some(30),
    )
    .await?;
    if response.status < 200 || response.status >= 300 {
        return Err(format!("Serper HTTP {}", response.status));
    }
    let data: serde_json::Value = serde_json::from_str(&response.body).map_err(|e| e.to_string())?;
    let mut out = Vec::new();
    if let Some(items) = data.get("organic").and_then(|v| v.as_array()) {
        for item in items.iter().take(limit) {
            out.push(WebSearchItem {
                title: item.get("title").and_then(|v| v.as_str()).unwrap_or("").into(),
                url: item.get("link").and_then(|v| v.as_str()).unwrap_or("").into(),
                snippet: item.get("snippet").and_then(|v| v.as_str()).unwrap_or("").into(),
            });
        }
    }
    Ok(out)
}

async fn search_tavily(api_key: &str, query: &str, limit: usize) -> Result<Vec<WebSearchItem>, String> {
    if api_key.is_empty() {
        return Err("Tavily 需要 API Key".into());
    }
    let body = serde_json::json!({
        "api_key": api_key,
        "query": query,
        "max_results": limit,
        "search_depth": "basic"
    });
    let response = crate::proxy::send_http(
        "POST",
        "https://api.tavily.com/search",
        Some(&[("Content-Type", "application/json")]),
        Some(body.to_string()),
        Some(30),
    )
    .await?;
    if response.status < 200 || response.status >= 300 {
        return Err(format!("Tavily HTTP {}", response.status));
    }
    let data: serde_json::Value = serde_json::from_str(&response.body).map_err(|e| e.to_string())?;
    let mut out = Vec::new();
    if let Some(items) = data.get("results").and_then(|v| v.as_array()) {
        for item in items.iter().take(limit) {
            out.push(WebSearchItem {
                title: item.get("title").and_then(|v| v.as_str()).unwrap_or("").into(),
                url: item.get("url").and_then(|v| v.as_str()).unwrap_or("").into(),
                snippet: item.get("content").and_then(|v| v.as_str()).unwrap_or("").into(),
            });
        }
    }
    Ok(out)
}

async fn search_duckduckgo(query: &str, limit: usize) -> Result<Vec<WebSearchItem>, String> {
    let url = format!(
        "https://api.duckduckgo.com/?q={}&format=json&no_html=1&skip_disambig=1",
        urlencoding::encode(query)
    );
    let response = crate::proxy::send_http("GET", &url, None, None, Some(30)).await?;
    if response.status < 200 || response.status >= 300 {
        return Ok(vec![]);
    }
    let data: serde_json::Value = serde_json::from_str(&response.body).map_err(|e| e.to_string())?;
    let mut out = Vec::new();
    if let (Some(text), Some(link)) = (
        data.get("AbstractText").and_then(|v| v.as_str()),
        data.get("AbstractURL").and_then(|v| v.as_str()),
    ) {
        out.push(WebSearchItem {
            title: data
                .get("Heading")
                .and_then(|v| v.as_str())
                .unwrap_or(query)
                .into(),
            url: link.into(),
            snippet: text.into(),
        });
    }
    if let Some(topics) = data.get("RelatedTopics").and_then(|v| v.as_array()) {
        for topic in topics {
            if out.len() >= limit {
                break;
            }
            if let (Some(text), Some(url)) = (
                topic.get("Text").and_then(|v| v.as_str()),
                topic.get("FirstURL").and_then(|v| v.as_str()),
            ) {
                out.push(WebSearchItem {
                    title: text.chars().take(80).collect(),
                    url: url.into(),
                    snippet: text.into(),
                });
            }
            if let Some(subs) = topic.get("Topics").and_then(|v| v.as_array()) {
                for sub in subs {
                    if out.len() >= limit {
                        break;
                    }
                    if let (Some(text), Some(url)) = (
                        sub.get("Text").and_then(|v| v.as_str()),
                        sub.get("FirstURL").and_then(|v| v.as_str()),
                    ) {
                        out.push(WebSearchItem {
                            title: text.chars().take(80).collect(),
                            url: url.into(),
                            snippet: text.into(),
                        });
                    }
                }
            }
        }
    }
    Ok(out.into_iter().take(limit).collect())
}
